reasoner_pydantic/base_model.py

```diff
@@ -0,0 +1,48 @@
+from typing import Callable
+import weakref
+from pydantic import BaseModel as PydanticBaseModel, PrivateAttr
+
+class BaseModel(PydanticBaseModel):
+    """
+    Custom base model for all classes
+
+    This provides a hash function that assumes all fields are either:
+
+    1. Immutable
+    2. Derived from BaseModel
+    3. Able to call a hash invalidation hook on their own
+    """
+
+    _hash: int = PrivateAttr(default=None)
+    _invalidate_hook: Callable = PrivateAttr(default=None)
+
+    def __hash__(self) -> int:
+        """ Hash function based on Pydantic implementation """
+        if not self._hash:
+            self._hash = hash(self.__class__) + hash(tuple(self.__dict__.values()))
```

Comment:
This is unstable in python <=3.5 since __dict__ is not guaranteed to be ordered. https://www.python.org/dev/peps/pep-0520/. But we do not want to sort __dict__ and introduce O(n log n) time complexity to the hashing function. It may be overkill to replace the __dict__ explicitly with an ordered dict like in https://stackoverflow.com/questions/27941581/replacing-default-dict-for-object-with-ordereddict. We should just note somewhere that this is only compatibly with python >= 3.6

Reply:
The setup.py file has `python_requires >= 3.6` so I think we should be good. Let me know if you think I should add a note somewhere else.